# ***STANDARD DEVIATION, VARIANCE ACTIVITY***

In [9]:
import pandas as pd
import plotly.express as px
import plotly.io as pio
import numpy as np 
pio.renderers.default = "iframe"
import requests
from io import StringIO
url = "https://www.openintro.org/data/csv/epa2012.csv"
headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3'}
response = requests.get(url, headers=headers)
response.raise_for_status() # Raise an exception for bad status codes
df = pd.read_csv(StringIO(response.text))

In [10]:
df

,model_yr,mfr_name,division,carline,mfr_code,model_type_index,engine_displacement,no_cylinders,transmission_speed,city_mpg,...,trans_lockup,trans_creeper_gear,drive_sys,drive_desc,fuel_usage,fuel_usage_desc,class,car_truck,release_date,fuel_cell
0,2012,aston martin,Aston Martin Lagonda Ltd,V12 Vantage,ASX,8,5.9,12.0,Manual(M6),11,...,N,N,R,"2-Wheel Drive, Rear",GP,Gasoline (Premium Unleaded Recommended),Two Seaters,car,2011-05-06,N
1,2012,aston martin,Aston Martin Lagonda Ltd,V8 Vantage,ASX,2,4.7,8.0,Auto(AM6),14,...,N,N,R,"2-Wheel Drive, Rear",GP,Gasoline (Premium Unleaded Recommended),Two Seaters,car,2011-07-11,N
2,2012,aston martin,Aston Martin Lagonda Ltd,V8 Vantage,ASX,11,4.7,8.0,Auto(AM7),14,...,N,N,R,"2-Wheel Drive, Rear",GP,Gasoline (Premium Unleaded Recommended),Two Seaters,car,2012-01-09,N
3,2012,aston martin,Aston Martin Lagonda Ltd,V8 Vantage,ASX,1,4.7,8.0,Manual(M6),13,...,N,N,R,"2-Wheel Drive, Rear",GP,Gasoline (Premium Unleaded Recommended),Two Seaters,car,2011-07-11,N
4,2012,aston martin,Aston Martin Lagonda Ltd,V8 Vantage S,ASX,3,4.7,8.0,Auto(AM7),14,...,N,N,R,"2-Wheel Drive, Rear",GP,Gasoline (Premium Unleaded Recommended),Two Seaters,car,2011-07-11,N
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1124,2012,Mercedes-Benz,Mercedes-Benz,MAYBACH 57,MBX,240,5.5,12.0,Auto(A5),10,...,Y,N,R,"2-Wheel Drive, Rear",GPR,Gasoline (Premium Unleaded Required),Large Cars,car,2011-11-24,NaN
1125,2012,Mercedes-Benz,Mercedes-Benz,MAYBACH 57 S,MBX,250,5.5,12.0,Auto(A5),10,...,Y,N,R,"2-Wheel Drive, Rear",GPR,Gasoline (Premium Unleaded Required),Large Cars,car,2011-11-24,NaN
1126,2012,Mercedes-Benz,Mercedes-Benz,MAYBACH 62,MBX,245,5.5,12.0,Auto(A5),10,...,Y,N,R,"2-Wheel Drive, Rear",GPR,Gasoline (Premium Unleaded Required),Large Cars,car,2011-12-01,NaN
1127,2012,Mercedes-Benz,Mercedes-Benz,MAYBACH 62 S,MBX,255,5.5,12.0,Auto(A5),10,...,Y,N,R,"2-Wheel Drive, Rear",GPR,Gasoline (Premium Unleaded Required),Large Cars,car,2011-11-24,NaN


### - one numerical column of choice is the combined MPG of vehicles.

In [24]:
mpg = df.iloc[:, 11]

In [28]:
print(mpg.describe())
print(mpg.median())

count    1129.000000
mean       21.906997
std         7.410413
min        10.000000
25%        18.000000
50%        21.000000
75%        25.000000
max       112.000000
Name: comb_mpg, dtype: float64
21.0


# what can we infer from the mean ?

### based on the statistic of the data we can see that the mean of the data is around 21.9 miles per gallon. this means that on average every car has around 21.9 miles per gallon. meanign that if we were to pick a car at random from the datasets, we would likely to get a MPG value close to 21.9

### the median tells us that the mean is slightly higher then median, implying that there are cars that has MPG way over the average range, which slightly pulls the mean up.

In [26]:
fig = px.histogram(mpg)
fig.show()

### from the plot we can see that the data is skewed towards the left, trailing off further on the right. mos tof the values MPG are focuses near the 0-40 range, with some that has exceptional MPG up to the 100s.

In [23]:
fig = px.box(mpg)
fig.update_layout(width = 300, height =500, title = "Combined MPG", yaxis_title="MPG Range")
fig.update_yaxes(range=[0,60])
fig.show()

### from the box plot, we can see the upper fence range as 35 while the lower fence range as 10, meaning the majority number of car's MPG lies within the 35-10 range, any cars that has MPG outside these values are considered outliers. we can also infer the interquartile range, in which the majority half lies within 25 to 18 MPG. 

# ***PRACTICE QUESTIONS***

## why can't we simple average deviation from the mean ?
- we can't simply average deviation from the mean because there would be positives and negatives deviation values, in which the sum total would be zeros. so we square them to avoid negative values. we uses square instead of taking the absolute values to be able to simplifies the math and make the curve of those values differentiable.
## why are deviations squared ?
- deviations are squared to avoid negatives values, which if we add normaly would sumup to zero.
## why do we divide by n-1 for a sample ?
- because sample mean are derived biasly from the sample, in which its deviation is smaller then the actual population mean. the mean of the sample is not an independent external reference, and it must follow the constraint in which the values up until n-1 are free while the last value is locked in to make the sum equals zero, losing one degree of freedom. therefore we remove the last value to remove the bias.
## can SD ever be  negative ?
- no, SD can not be negative because it is the square root of the squared values of the deviations. so it is mathematically not possible to be negative.

## which dataset is more consistent: SD = 2 or SD = 15 ?
- SD = 2 is more consistent because it means the data is less spread out, in which on average the data values are only spread out by + - 2 instead of + - 15.

## how do outliers affect variance and standard deviation ?
- Standard Deviation are heavily sensitive to outliers, in which a few outliers can majorly influence the data.

## when is standard deviation more useful than variance ?
- standard deviation are usually always more useful then variance because they are in the original unit of the data beign measured, so its easier to compare and perhaps visually view the datasets given standard deviation.